packages/core/src/components/forms/useAsyncControllableValue.ts

```diff
@@ -0,0 +1,138 @@
+
+    // Whether we are in the middle of a composition event.
+    const [isComposing, setIsComposing] = React.useState(false);
+
+    // Whether there is a pending update we are expecting from a parent component.
+    const [hasPendingUpdate, setHasPendingUpdate] = React.useState(false);
+
+    const cancelPendingCompositionEnd = React.useRef<(() => void) | null>(null);
+
+    const handleCompositionStart: React.CompositionEventHandler<E> = React.useCallback(
+        event => {
+            cancelPendingCompositionEnd.current?.();
+            setIsComposing(true);
+            onCompositionStart?.(event);
+        },
+        [onCompositionStart],
+    );
+
+    // creates a timeout which will set `isComposing` to false after a delay
+    // returns a function which will cancel the timeout if called before it fires
+    const createOnCancelPendingCompositionEnd = React.useCallback(() => {
+        const timeoutId = window.setTimeout(
+            () => setIsComposing(false),
+            ASYNC_CONTROLLABLE_VALUE_COMPOSITION_END_DELAY,
+        );
+        return () => window.clearTimeout(timeoutId);
+    }, []);
+
+    const handleCompositionEnd: React.CompositionEventHandler<E> = React.useCallback(
+        event => {
+            // In some non-latin languages, a keystroke can end a composition event and immediately afterwards start another.
+            // This can lead to unexpected characters showing up in the text input. In order to circumvent this problem, we
+            // use a timeout which creates a delay which merges the two composition events, creating a more natural and predictable UX.
+            // `this.state.nextValue` will become "locked" (it cannot be overwritten by the `value` prop) until a delay (10ms) has
+            // passed without a new composition event starting.
+            cancelPendingCompositionEnd.current = createOnCancelPendingCompositionEnd();
+            onCompositionEnd?.(event);
+        },
+        [createOnCancelPendingCompositionEnd, onCompositionEnd],
+    );
+
+    const handleChange: React.ChangeEventHandler<E> = React.useCallback(
+        event => {
+            const { value: targetValue } = event.target;
+            setNextValue(targetValue);
+            onChange?.(event);
+        },
+        [onChange],
+    );
+
+    if (isComposing || propValue === undefined) {
+        // don't derive anything from props if:
+        // - in uncontrolled mode, OR
+        // - currently composing, since we'll do that after composition ends
+    } else {
+        const userTriggeredUpdate = nextValue !== value;
+
+        if (userTriggeredUpdate) {
+            if (propValue === nextValue) {
+                // parent has processed and accepted our update
+                if (hasPendingUpdate) {
+                    setValue(propValue);
+                    setHasPendingUpdate(false);
+                } else if (value !== nextValue) {
+                    setValue(nextValue);
+                }
```

Comment:
Taking this a bit further, I think the nested logic can be much flatter.

```tsx
if (!isStateDerivationPaused) {
    const userTriggeredUpdate = nextValue !== value;

    if (userTriggeredUpdate && propValue === nextValue) {
        // parent has processed and accepted our update
        setValue(propValue);
        setHasPendingUpdate(false);
    } else if (userTriggeredUpdate && propValue === value) {
        // we have sent the update to our parent, but it has not been processed yet. just wait.
        // DO NOT set nextValue here, since that will temporarily render a potentially stale controlled value,
        // causing the cursor to jump once the new value is accepted
        if (!hasPendingUpdate) {
            setHasPendingUpdate(true);
        }
    } else if (userTriggeredUpdate || value !== propValue || hasPendingUpdate) {
        // accept controlled update, could be confirming or denying user action
        setValue(propValue);
        setNextValue(propValue);
        setHasPendingUpdate(false);
    }
}
```

But I could definitely use help making sure I refactored that correctly. The general strategy was to propagate the `userTriggeredUpdate` condition down and then remove layers of nesting.

If you would rather not think about further refactors and prefer to keep the existing structure as-is to match the current `getDerivedStateFromProps` approach, I think that's okay too!

Reply:
I refactored this bit to be flatter and (arguably) more readable than the initial `getDerivedStateFromProps` version in the latest commit. Thanks for the ideas above!